src/sync/app.cpp

```diff
@@ -16,67 +16,71 @@
 //
 ////////////////////////////////////////////////////////////////////////////
 
-#include "app.hpp"
-#include <sstream>
-#include "sync_manager.hpp"
-#include "generic_network_transport.hpp"
-#include "app_credentials.hpp"
-#include "../external/json/json.hpp"
+#include "sync/app.hpp"
 
-// wrap an optional json key into the Optional type
-#define WRAP_JSON_OPT(JSON, KEY, RET_TYPE) \
-JSON.find(KEY) != JSON.end() ? Optional<RET_TYPE>(JSON[KEY].get<RET_TYPE>()) : realm::util::none
+#include "sync/app_credentials.hpp"
+#include "sync/generic_network_transport.hpp"
+#include "sync/sync_manager.hpp"
+
+#include <json.hpp>
+#include <sstream>
 
 namespace realm {
 namespace app {
 
-std::shared_ptr<App> App::app(const std::string app_id,
-                              const realm::util::Optional<App::Config> config)
+// wrap an optional json key into the Optional type
+template <typename T>
+Optional<T> get_optional(const nlohmann::json& json, const std::string& key)
 {
-    return std::make_shared<App>(app_id, config);
+    auto it = json.find(key);
+    return it != json.end() ? Optional<T>(it->get<T>()) : realm::util::none;
 }
 
-const std::string App::default_base_url = "https://stitch.mongodb.com";
-const std::string App::base_path = "/api/client/v2.0";
-const std::string App::app_path = "/app";
-const std::string App::auth_path = "/auth";
-const uint64_t App::default_timeout_ms = 60000;
+const static std::string default_base_url = "https://stitch.mongodb.com";
+const static std::string default_base_path = "/api/client/v2.0";
+const static std::string default_app_path = "/app";
+const static std::string default_auth_path = "/auth";
+const static uint64_t    default_timeout_ms = 60000;
+
+App::App(const Config& config)
+: m_config(config)
+, m_base_route(config.base_url.value_or(default_base_url) + default_base_path)
+, m_app_route(m_base_route + default_app_path + "/" + config.app_id)
+, m_auth_route(m_app_route + default_auth_path)
+, m_request_timeout_ms(config.default_request_timeout_ms.value_or(default_timeout_ms))
+{
+    REALM_ASSERT(m_config.transport_generator);
+}
 
-static std::unique_ptr<error::AppError> handle_error(const Response response) {
+static Optional<AppError> handle_error(const Response& response) {
```

Comment:
I don't believe this should be optional. The idea is to parse the error as best we can, but if for some reason we cannot, we still need to let the binding know that _some_ error has occurred, even if we return "Unknown" or -1.

Reply:
I combined the error check and error generation together so this whole mechanism is more robust.